### ISSUE TITLE
feature request: a strict strict mode

### ISSUE THREAD

**Issue by sam-github** [eslint/eslint#3010]
strict mode is getting more lax as more features are defined, I particularly run into this when strict mode  code that is valid in io.js throws exceptions in node v0.10, it would be nice to have this caught by eslint in dev, rather than CI.

Perhaps a 'strict-strict' mode that checks for a sub-set of strict that was valid in node v0.10 (that is, whatever ES version it's V8 supported) would be helpful?

**Comment by ilyavolodin** [comment]
Can you provide code example that's causing this issue, version of ESLint your are running and configuration for `strict` rule you are using?

**Comment by sam-github** [comment]
eslintrc: https://gist.github.com/6c9095c20b15ca4285a3

fu.js, with a function inside a conditional (the use of process is to have something to put in the if, and to use the function - eslint warns on unused functions, as it should)
```
'use strict';

if (process.exit) {
  process.on('exit', hello);
  function hello() {}
}
```

eslint warnings: none

node v0.10:
```
% nvm use 0.10
Now using node v0.10.38
sn/strong-pm (configurable-app-baseport $%) % node fu.js

fu.js:5
  function hello() {}
  ^^^^^^^^
SyntaxError: In strict mode code, functions can only be declared at top level or immediately within another function.
    at Module._compile (module.js:439:25)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Function.Module.runMain (module.js:497:10)
    at startup (node.js:119:16)
    at node.js:935:3
```

iojs 2.2.1: no error, under recent js updates, the usage in fu.js is well-defined

**Comment by mysticatea** [comment]
It looks not related with ESLint.

> SyntaxError: In strict mode code, functions can only be declared at top level or immediately within another function.

**Comment by mysticatea** [comment]
Sorry, hm, did not ESLint have a rule to warn function declarations in `if`/`while`?

**Comment by IanVS** [comment]
You have the strict rule turned off.  Try turning it on and setting it to `"global"`.

https://gist.github.com/sam-github/6c9095c20b15ca4285a3#file-eslintrc-L123

**Comment by btmills** [comment]
Is [block-scoped-var](http://eslint.org/docs/rules/no-inner-declarations) what you're looking for?

**Comment by nzakas** [comment]
The issue is that function declarations were not allowed to be in blocks in ES5 but are allowed in ES6. We probably need a no-block-func rule for this.

@sam-github we can't add a rule that implements all of strict mode, but we can certainly add does for the posts y we are missing (we already have a lot).

**Comment by btmills** [comment]
I just realized my earlier comment linked to no-inner-declarations, but I typed out block-scoped-var. Fixed, no-inner-declarations is what I meant!

**Comment by sam-github** [comment]
@nzakas all of strict mode would be too much, I agree, but a `strict-es5-compat` would be more reasonable: it would only have to include the ***differences*** in strict mode from es5 to es6. The only things I've run into are variable scope changes.

**Comment by sam-github** [comment]
@btmills the problem with `no-inner-declarations` is that it applies to non-strict mode too.

Inner declarations work fine in our code outside of strict mode (which is node/v8-only, I'm not concerned about browser compatibility), I don't want to be warned about valid code, its only in es5 strict mode that v8 errors out on inner declarations.

As we slowly introduce strict mode through our code base, we make any changes necessary. But at the moment, I need to run node v0.10 (or wait for CI to do so) before I discover all the strict-mode violations, rather than getting told be eslint or io.js (which is es6-ish).

**Comment by nzakas** [comment]
@sam-github we don't bundle that much functionality into a single rule, nor can we enable/disable rules based on strict mode. It would be up to you to apply `no-inner-declarations` where necessary.

As an aside, we should probably reorganize the rules list so those that mimic strict mode restrictions are in one spot.

**Comment by sam-github** [comment]
@nzakas I thought you had rule categories that could be imported as a whole, such as `extends: "es5-compatible"`, but that's OK, I'll use any combination of eslint config that gets the job done.

But to be clear: `no-inner-declarations`, as-is, can't be used for our code, because it applies to non-strict-mode functions, as well as strict, and thus will warn about valid non-strict code.

We are gradually introducing strict mode, file by file, as we are able.

So, I could use a flag on that `no-inner-declarations`, or a new rule like `es5-strict-mode-inner-declarations`, that warns about usage of inner declarations that are invalid in es5 strict mode.

**Comment by btmills** [comment]
@sam-github are you able to add a comment to the top of your files as you convert them? Would having `no-inner-declarations` off in your `.eslintrc`, then enabling it in converted files with `/* eslint no-inner-declarations: 2 */` work?

**Comment by nzakas** [comment]
@sam-github we do have shareable configs, but we only ship with one for back compat. You are free to create your own that you could use in that way.

What you're asking for is more fine-grained than we make our rules, so I'd suggest making a custom rule to cover this case.

**Comment by sam-github** [comment]
@btmills I could add an inline eslint rule in every file we do `'use strict';`, but that's too manual, I want my tools to tell me when my code is invalid in strict-mode.

That's OK, we don't rely on eslint telling us when our js is invalid, our CI still does that, albeit after the PR is made, thanks for considering this.
